src/process_utility.c

```diff
@@ -695,32 +691,58 @@ process_altertable_change_owner(Hypertable *ht, AlterTableCmd *cmd)
 }
 
 static void
+process_add_constraint_chunk(Hypertable *ht, Oid chunk_relid, void *arg)
+{
+	Oid			hypertable_constraint_oid = *((Oid *) arg);
+	Chunk	   *chunk = chunk_get_by_relid(chunk_relid, ht->space->num_dimensions, true);
+
+	chunk_constraint_create_on_chunk(ht, chunk, hypertable_constraint_oid);
+}
+
+static void
 process_altertable_add_constraint(Hypertable *ht, const char *constraint_name)
 {
-	CatalogSecurityContext sec_ctx;
+	Oid			hypertable_constraint_oid = get_relation_constraint_oid(ht->main_table_relid, constraint_name, false);
 
 	Assert(constraint_name != NULL);
 
-	catalog_become_owner(catalog_get(), &sec_ctx);
-	process_utility_set_expect_chunk_modification(true);
-	process_add_hypertable_constraint(ht, constraint_name);
-	process_utility_set_expect_chunk_modification(false);
-	catalog_restore_user(&sec_ctx);
+	foreach_chunk(ht, process_add_constraint_chunk, &hypertable_constraint_oid);
 }
 
 static void
+process_drop_constraint_index_chunk(Hypertable *ht, Oid chunk_relid, void *arg)
+{
+	Oid			hypertable_constraint_index_oid = *((Oid *) arg);
+	Chunk	   *chunk = chunk_get_by_relid(chunk_relid, ht->space->num_dimensions, true);
+
+	chunk_index_delete(chunk, hypertable_constraint_index_oid, false);
+}
+
+
+static void
 process_altertable_drop_constraint(Hypertable *ht, AlterTableCmd *cmd)
 {
 	char	   *constraint_name = NULL;
 	CatalogSecurityContext sec_ctx;
+	Oid			hypertable_constraint_oid,
+				hypertable_constraint_index_oid;
 
 	constraint_name = cmd->name;
 	Assert(constraint_name != NULL);
 
+	hypertable_constraint_oid = get_relation_constraint_oid(ht->main_table_relid, constraint_name, false);
+	hypertable_constraint_index_oid = get_constraint_index(hypertable_constraint_oid);
+
 	catalog_become_owner(catalog_get(), &sec_ctx);
+
 	process_utility_set_expect_chunk_modification(true);
 	process_drop_hypertable_constraint(ht, constraint_name);
 	process_utility_set_expect_chunk_modification(false);
+
+	if (OidIsValid(hypertable_constraint_index_oid))
+	{
+		foreach_chunk(ht, process_drop_constraint_index_chunk, &hypertable_constraint_index_oid);
```

Comment:
Can't we call `chunk_index_delete_children_of(hypertable_constraint_index_oid)` here instead of looping for every chunk?

Reply:
Ideally, we'd use this loop to call `RemoveConstraintById()` on all chunk constraints.

Reply:
fixed first comment; leaving second for another PR